ships/0025-event-driven-builds.md

```diff
@@ -0,0 +1,300 @@
+<!--
+Copyright The Shipwright Contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+---
+title: event-driven-builds
+authors:
+  - "@sbose787"
+  
+reviewers:
+  - "@gmontero"
+  - "@adamkaplan"
+  - "@ImJasonH"
+  - "@SaschaSchwarze0"
+  - "@HeavyWombat"
+  
+approvers:
+  - "@adamkaplan"
+  - "@SaschaSchwarze0"
+
+
+ 
+ 
+creation-date: 2021-11-03
+status: implementable
+
+---
+
+# Git Event-driven triggering of Shipwright Builds
+
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Open Questions [optional]
+
+1. Almost every user would need an exposed `Service`, how do we create a vendor-agnostic `Ingress` object ?
+
+
+## Summary
+
+Trigger the execution of an image build based on a commit/push event from a relevant source code repository. 
+
+
+
+## Motivation
+
+This enhancement proposal aims to provide an API to enable users to express the intent of having their builds triggered by events from a Git Repository.
+
+
+### Goals
+
+* Build a technology-agostic user experience for Git-triggered build executions.
+* Add support for 'reacting' to events from Github and Gitlab repositories.
+
+
+
+### Non-Goals
+
+* Support events from generic git servers. The implementation should be extensible to support those in a non-breaking manner.
+* Creation of the vendor-specific Ingress resources to expose the webhook URL.
+
+
+## Proposal
+
+
+### User Stories [optional]
+
+Detail the things that people will be able to do if this is implemented. Include as much detail as
+possible so that people can understand the "how" of the system. The goal here is to make this feel
+real for users without getting bogged down.
+
+#### Story 1
+As a user, I would like to define a `Build` and trigger the execution of the same upon pushes to my Git repository.
+
+#### Story 2
+As a user, I would like to define a `Build` and the execution of the same for pushes to any branch in that Git repository.
+
+#### Story 3
+As a user, I would like to configure a secure webhook URL for triggering `Builds`.
+
+
+### Implementation Notes
+
+Upon specification of the **new** API field `.spec.webhook`, the Shipwright Build Controller will do the needful to generate a webhook URL and 
+provide the information on the same in the `status` of the `Build` resource.
+
+
+```
+spec:
+  ...
+  ...
+  webhook:
+    type: github
+    imageTagPolicy: short_sha    # optional, allowed values: 'short_sha' , 'branch'. Defaults to 'branch'.
+    secretRef:                   # optional, will be genereated if not specified.
+      name: my-webhook-secret.  
```

Comment:
I wonder if optional is reasonable here because if we generate it, then the user will need to decode it to get the webhook token to then store it in GitHub, right?

Reply:
Correct, this is more of a knob of convenience :)